@RequestParam required true

### Describe the bug

Like the title says the problem is into annotation `@RequestParam(required=true)`

Into spring you can specify if a request param is required or not and spring ( if required is set ) and it's not passed raise a BadRequestException.


But in quarkus i've this problematic point:

Required param seems it's not supported each time I added this the app gave me this exception:
```java

Caused by: java.lang.IllegalArgumentException: Using required @RequestMapping is not supported. Offending method is 'it.test.controller.TestController#customException'
	at io.quarkus.spring.web.resteasy.reactive.deployment.SpringWebResteasyReactiveProcessor$1.transform(SpringWebResteasyReactiveProcessor.java:312)
	at org.jboss.resteasy.reactive.common.processor.transformation.AnnotationsTransformer.apply(AnnotationsTransformer.java:74)
	at org.jboss.jandex.AnnotationOverlayImpl$2.apply(AnnotationOverlayImpl.java:253)
	at org.jboss.jandex.AnnotationOverlayImpl$2.apply(AnnotationOverlayImpl.java:246)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.jboss.jandex.AnnotationOverlayImpl.getAnnotationsFor(AnnotationOverlayImpl.java:246)
	at org.jboss.jandex.AnnotationOverlayImpl.annotation(AnnotationOverlayImpl.java:129)
	at org.jboss.resteasy.reactive.common.processor.transformation.AnnotationStore.getAnnotation(AnnotationStore.java:52)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.collectEndpoints(EndpointIndexer.java:480)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:403)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:312)
	... 11 more
```
 

So i need to validate each request param field manually?


Another problem is the default value, if my request param is a boolean, and i don't pass it from the caller, Quarkus set a default value. Why?

It's totally wrong..

IE:

<img width="978" height="389" alt="Image" src="https://github.com/user-attachments/assets/46a71261-3ec1-4144-acbb-d1c96f11ecb2" />

In this example if you don't pass firstException param Quarkus set it to false.. but does not make sense because the param should be NULL so raise an exception..

### Expected behavior

What I expect is that, since the RequestParam annotation is supported, its functions such as required and raising an exception if a required parameter is not passed are also supported.

### Actual behavior

_No response_

### How to Reproduce?

Download my project attached.

Run with `./mvnw quarkus:dev`


Try to call the endpoint `localhost:8080/test/exception` without passing param `firstException` and can u see the default value problem

Try to add `required=true` into the TestController under `@RequestParam` annotation e you can see the error during startup

[quarkus-request-param.zip](https://github.com/user-attachments/files/22597679/quarkus-request-param.zip)

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21

### Quarkus version or git rev

3.26.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvnw quarkus:dev

### Additional information

_No response_

> > Hi [@aureamunoz](https://github.com/aureamunoz) Thank you for your reply, but will this trigger an exception that can be caught in the global exception handler? Is this `WebApplicationException` ?
> > As for the possibility of defining a `@RequestParam` (Spring annotation) as required or not required for an endpoint, is this feasible?
> 
> Hi [@GuidoFarport](https://github.com/GuidoFarport) Yes, the validation throws a WebApplicationException, which can be caught by your global exception handler. In the reproducer it’s handled by `ExceptionMappersSpring#unsafedException`.
> 
> Regarding the required attribute of [@RequestParam](https://github.com/RequestParam), it’s supported in Quarkus and works the same way as in Spring Web.

Hi @aureamunoz sorry but in which Quarkus version `@RequestParam` annotation works like Spring Web?


Because if you try to download the project i attached in this ticket and modify the TestController endpoint by adding the param `required=true`  you recive error when quarkus start..

(ie)

<img width="1031" height="385" alt="Image" src="https://github.com/user-attachments/assets/b900f314-40e0-4678-ab15-b1b00c3acafd" />


stacktrace error:

```java

Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Using required @RequestMapping is not supported. Offending method is 'it.test.controller.TestController#customException'
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:351)
	at io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor.setupEndpoints(ResteasyReactiveProcessor.java:710)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:874)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)

```

--- a/extensions/spring-web/resteasy-reactive/runtime/src/main/java/io/quarkus/spring/web/resteasy/reactive/runtime/SpringRequestParamHandler.java
+++ b/extensions/spring-web/resteasy-reactive/runtime/src/main/java/io/quarkus/spring/web/resteasy/reactive/runtime/SpringRequestParamHandler.java
@@ -4,7 +4,9 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-import org.jboss.resteasy.reactive.common.jaxrs.ResponseImpl;
+import jakarta.ws.rs.WebApplicationException;
+import jakarta.ws.rs.core.Response;
+
 import org.jboss.resteasy.reactive.common.model.ResourceClass;
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.jboss.resteasy.reactive.server.model.HandlerChainCustomizer;
@@ -21,18 +23,19 @@ public class SpringRequestParamHandler implements HandlerChainCustomizer {
     @Override
     public List<ServerRestHandler> handlers(HandlerChainCustomizer.Phase phase, ResourceClass resourceClass,
             ServerResourceMethod resourceMethod) {
-        if (phase == Phase.AFTER_RESPONSE_CREATED) {
+        if (phase == Phase.RESOLVE_METHOD_PARAMETERS) {
             return Collections.singletonList(new ServerRestHandler() {
                 @Override
                 public void handle(ResteasyReactiveRequestContext requestContext) throws Exception {
                     Map<String, List<String>> parametersMap = requestContext.serverRequest().getQueryParamsMap();
                     if (parametersMap.isEmpty()) {
-                        ResponseImpl response = (ResponseImpl) requestContext.getResponse().get();
-                        response.setStatus(400);
+                        throw new WebApplicationException("Missing required param in method '" + resourceMethod.getName() + "'",
+                                Response.Status.BAD_REQUEST);
                     }
                 }
             });
         }
         return Collections.emptyList();
     }
+
 }

--- a/extensions/spring-web/resteasy-reactive/tests/src/test/java/io/quarkus/spring/web/requestparam/RequestParamController.java
+++ b/extensions/spring-web/resteasy-reactive/tests/src/test/java/io/quarkus/spring/web/requestparam/RequestParamController.java
@@ -5,13 +5,19 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+import jakarta.ws.rs.WebApplicationException;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.util.MultiValueMap;
+import org.springframework.web.bind.annotation.ExceptionHandler;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
 
 @RestController
 @RequestMapping
@@ -67,6 +73,20 @@ public class RequestParamController {
             result = "Parameters are " + entry.getKey() + "=" + entry.getValue().stream().collect(Collectors.joining(", "));
         }
         return result;
+    }
+
+    @GetMapping("/api/foos/paramRequired")
+    public String getFoosParamRequired(@RequestParam String id) {
+        throw new IllegalStateException("Unexpected state. Should have not been called");
+    }
+
+    @RestControllerAdvice
+    public static class RestExceptionHandler {
+
+        @ExceptionHandler(WebApplicationException.class)
+        public ResponseEntity<Object> handleException(Exception ex) {
+            return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
+        }
     }
 
 }

--- a/extensions/spring-web/resteasy-reactive/tests/src/test/java/io/quarkus/spring/web/requestparam/RequestParamControllerTest.java
+++ b/extensions/spring-web/resteasy-reactive/tests/src/test/java/io/quarkus/spring/web/requestparam/RequestParamControllerTest.java
@@ -26,7 +26,7 @@ public class RequestParamControllerTest {
         // In Spring, method parameters annotated with @RequestParam are required by default.
         when().get("/api/foos")
                 .then()
-                .statusCode(400);
+                .statusCode(400).body(containsString("Missing required param in method"));
     }
 
     @Test
@@ -121,6 +121,14 @@ public class RequestParamControllerTest {
         when().post("/api/foos/multiMap")
                 .then()
                 .statusCode(400);
+    }
+
+    @Test
+    public void shouldFailWithBadRequestWhenMissingMandatoryParam() throws Exception {
+        // In Spring, method parameters annotated with @RequestParam are required by default.
+        when().get("/api/foos/paramRequired")
+                .then()
+                .statusCode(400).body(containsString("Missing required param in method"));
     }
 
 }

